### ISSUE TITLE
Checksum doesn't match: corrupted data

### ISSUE THREAD

**Issue by RonanMorgan** [ClickHouse/ClickHouse#18135]
Hello everyone,

I have the following errors on one of my clickhouse servers :

```
2020.12.16 10:23:20.590582 [ 2938445 ] {} <Error> void DB::BackgroundProcessingPool::workLoopFunc(): Code: 40, e.displayText() = DB::Exception: Checksum doesn't match: corrupted data. Reference: 3981c13d9526fa31ccb8afc1b4bc00c4. Actual: a2ef2b1f41451cb488947d435e967be2. Size of compressed block: 155036. The mismatch is caused by single bit flip in data block at byte 105203, bit 1. This is most likely due to hardware failure. If you receive broken data over network and the error does not repeat every time, this can be caused by bad RAM on network interface controller or bad controller itself or bad RAM on network switches or bad CPU on network switches (look at the logs on related network switches; note that TCP checksums don't help) or bad RAM on host (look at dmesg or kern.log for enormous amount of EDAC errors, ECC-related reports, Machine Check Exceptions, mcelog; note that ECC memory can fail if the number of errors is huge) or bad CPU on host. If you read data from disk, this can be caused by disk bit rott. This exception protects ClickHouse from data corruption due to hardware failures.: (while reading column LANDINGPAGEURL): While executing MergeTreeSequentialSource, Stack trace (when copying this message, always include the lines below):

0. DB::CompressedReadBufferBase::readCompressedData(unsigned long&, unsigned long&) @ 0xd39afbb in /usr/bin/clickhouse
1. DB::CompressedReadBufferFromFile::nextImpl() @ 0xdafae28 in /usr/bin/clickhouse
2. ? @ 0xd51bd47 in /usr/bin/clickhouse
3. DB::DataTypeString::deserializeBinaryBulk(DB::IColumn&, DB::ReadBuffer&, unsigned long, double) const @ 0xd51b17b in /usr/bin/clickhouse
4. DB::MergeTreeReaderCompact::readData(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::IColumn&, DB::IDataType const&, unsigned long, unsigned long, unsigned long, bool) @ 0xe12c77f in /usr/bin/clickhouse
5. DB::MergeTreeReaderCompact::readRows(unsigned long, bool, unsigned long, std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >&) @ 0xe12be77 in /usr/bin/clickhouse
6. DB::MergeTreeSequentialSource::generate() @ 0xe14d229 in /usr/bin/clickhouse
7. DB::ISource::work() @ 0xe2ded8a in /usr/bin/clickhouse
8. DB::SourceWithProgress::work() @ 0xe443a6a in /usr/bin/clickhouse
9. ? @ 0xe317b6c in /usr/bin/clickhouse
10. DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) @ 0xe314c97 in /usr/bin/clickhouse
11. DB::PipelineExecutor::executeStep(std::__1::atomic<bool>*) @ 0xe31397c in /usr/bin/clickhouse
12. DB::PullingPipelineExecutor::pull(DB::Chunk&) @ 0xe320429 in /usr/bin/clickhouse
13. DB::PullingPipelineExecutor::pull(DB::Block&) @ 0xe320500 in /usr/bin/clickhouse
14. DB::PipelineExecutingBlockInputStream::readImpl() @ 0xe30fc34 in /usr/bin/clickhouse
15. DB::IBlockInputStream::read() @ 0xd4655a5 in /usr/bin/clickhouse
16. DB::MergeTreeDataMergerMutator::mergePartsToTemporaryPart(DB::FutureMergedMutatedPart const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::BackgroundProcessListEntry<DB::MergeListElement, DB::MergeInfo>&, std::__1::shared_ptr<DB::RWLockImpl::LockHolderImpl>&, long, DB::Context const&, std::__1::unique_ptr<DB::IReservation, std::__1::default_delete<DB::IReservation> > const&, bool) @ 0xe076307 in /usr/bin/clickhouse
17. DB::StorageMergeTree::merge(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*) @ 0xde25119 in /usr/bin/clickhouse
18. DB::StorageMergeTree::mergeMutateTask() @ 0xde294e9 in /usr/bin/clickhouse
19. DB::BackgroundProcessingPool::workLoopFunc() @ 0xdfd7493 in /usr/bin/clickhouse
20. ? @ 0xdfd7fd1 in /usr/bin/clickhouse
21. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x7b1a93d in /usr/bin/clickhouse
22. ? @ 0x7b1e463 in /usr/bin/clickhouse
23. start_thread @ 0x82de in /usr/lib64/libpthread-2.28.so
24. __GI___clone @ 0xfbe83 in /usr/lib64/libc-2.28.so
 (version 20.11.5.18 (official build))
```

It seems to be due to a hardware or network issue, which will happen time to time I guess.

I don't want to deactivate the ckecksum or remove data from my table, do you have any advice about that ? 
Is there any way that clickhouse return an error to the client at save in order to retry the save for ex ?

Also, why does it happen in loop ? It mean that clickhouse is trying to use this data again and again but if there is no way to fix it I don't see the point ...

**Comment by g1rlp0w3r** [comment]
Hello,

I'm also having the same issue here. And also -not sure if it is related- I can't remove data from the failing block.
Anny suggestions? This problem start happening days before I've inserted the data.

`ClickHouse server version 21.2.5.5 (official build).`

`DB::Exception: Checksum doesn't match: corrupted data. Reference: 782dd83ec202639b1e5de5cb2540ad56. Actual: ccc788c8e8ec0bbacc36b6d2dd0ea5d5. Size of compressed block: 29375. The mismatch is caused by single bit flip in data block at byte 26483, bit 1. This is most likely due to hardware failure. If you receive broken data over network and the error does not repeat every time, this can be caused by bad RAM on network interface controller or bad controller itself or bad RAM on network switches or bad CPU on network switches (look at the logs on related network switches; note that TCP checksums don't help) or bad RAM on host (look at dmesg or kern.log for enormous amount of EDAC errors, ECC-related reports, Machine Check Exceptions, mcelog; note that ECC memory can fail if the number of errors is huge) or bad CPU on host. If you read data from disk, this can be caused by disk bit rott. This exception protects ClickHouse from data corruption due to hardware failures.: (while reading column licensor): (while reading from part /data/clickhouse/lib/clickhouse/data/charts/report_usage/20210301_2895_9969_5_15286/ from mark 412 with max_rows_to_read = 8192): While executing MergeTreeThread. Stack trace:

0. DB::CompressedReadBufferBase::readCompressedData(unsigned long&, unsigned long&, bool) @ 0xe5bfbee in /usr/bin/clickhouse
1. DB::CompressedReadBufferFromFile::nextImpl() @ 0xe5c0f80 in /usr/bin/clickhouse
2. ? @ 0xe75ce67 in /usr/bin/clickhouse
3. DB::DataTypeString::deserializeBinaryBulk(DB::IColumn&, DB::ReadBuffer&, unsigned long, double) const @ 0xe75c24b in /usr/bin/clickhouse
4. DB::IDataType::deserializeBinaryBulkWithMultipleStreams(COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, DB::IDataType::DeserializeBinaryBulkSettings&, std::__1::shared_ptr<DB::IDataType::DeserializeBinaryBulkState>&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > >*) const @ 0xe77508c in /usr/bin/clickhouse
5. DB::MergeTreeReaderWide::readData(DB::NameAndTypePair const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, bool, unsigned long, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > >&) @ 0xf42d602 in /usr/bin/clickhouse
6. DB::MergeTreeReaderWide::readRows(unsigned long, bool, unsigned long, std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >&) @ 0xf42c76e in /usr/bin/clickhouse
7. DB::MergeTreeRangeReader::DelayedStream::finalize(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >&) @ 0xf434c6a in /usr/bin/clickhouse
8. DB::MergeTreeRangeReader::continueReadingChain(DB::MergeTreeRangeReader::ReadResult&, unsigned long&) @ 0xf438efa in /usr/bin/clickhouse
9. DB::MergeTreeRangeReader::read(unsigned long, std::__1::deque<DB::MarkRange, std::__1::allocator<DB::MarkRange> >&) @ 0xf437ea6 in /usr/bin/clickhouse
10. DB::MergeTreeBaseSelectProcessor::readFromPartImpl() @ 0xf4321f3 in /usr/bin/clickhouse
11. DB::MergeTreeBaseSelectProcessor::readFromPart() @ 0xf432ecd in /usr/bin/clickhouse
12. DB::MergeTreeBaseSelectProcessor::generate() @ 0xf4316eb in /usr/bin/clickhouse
13. DB::ISource::tryGenerate() @ 0xf606e15 in /usr/bin/clickhouse
14. DB::ISource::work() @ 0xf606b2a in /usr/bin/clickhouse
15. DB::SourceWithProgress::work() @ 0xf761fca in /usr/bin/clickhouse
16. ? @ 0xf64172d in /usr/bin/clickhouse
17. DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) @ 0xf63e521 in /usr/bin/clickhouse
18. ? @ 0xf642f26 in /usr/bin/clickhouse
19. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x84f700f in /usr/bin/clickhouse
20. ? @ 0x84faaa3 in /usr/bin/clickhouse
21. start_thread @ 0x7fa3 in /usr/lib/x86_64-linux-gnu/libpthread-2.28.so
22. clone @ 0xf94cf in /usr/lib/x86_64-linux-gnu/libc-2.28.so`

**Comment by alexey-milovidov** [comment]
@RonanMorgan @g1rlp0w3r Your data is broken due to faulty hardware.

If you are using Replicated tables (redundancy), the data will be recovered automatically (no action required).

If you are not using Replicated tables, it is possible to continue operation by moving out broken data part manually (move the directory `/data/clickhouse/lib/clickhouse/data/charts/report_usage/20210301_2895_9969_5_15286/` somewhere and restart the server).

To summarize:
1. Your hardware is broken.
2. Replication is the only way to avoid data loss in presense of broken hardware.
3. ClickHouse allows you to detect broken hardware and ensures that silent data corruption is not possible.
